### PR TITLE
31-hide-spontaneous

### PR DESCRIPTION
**Pull request by tom08**
Create experience now hides attendance field unless type does not need verification

### CODE REVIEW

**Comment by aaaustin10** [comment]
I noticed that you can delete the welcome.js and we can finally have perfect coverage (yay). I also noticed that your missing one line of JavaScript coverage.

**Comment by aaaustin10** [comment]
Conclusion doesn't appear anywhere on the page ever that I can tell and that was one of the requirements of this issue.

**Comment by tom08** [comment]
I thought we had decided to make them go through the conclusion page anyway.

File: exdb/tests/integration_tests.py [hunk 1]
```diff
@@ -261,9 +274,10 @@ def test_valid_past_experience_creation_view_submit(self):
         data = self.get_post_data(start, end)
         data['attendance'] = 1
         data['type'] = self.test_past_type.pk
+        data['conclusion'] = "Test conclusion"
         self.clients['ra'].post(reverse('create_experience'), data)
         self.assertEqual('co', Experience.objects.get(name='test').status,
-                         "Experience should have been saved with completed status")
+                         "Experience should have been saved with approved status")
```
Comment (shaneallgeier) [inline exdb/tests/integration_tests.py]: This should probably be reverted back to `completed`.

**Comment by tom08** [comment]
Note to remember fixing the js formatting problems.

**Comment by enki211** [comment]
Looks good

File: exdb/forms.py [hunk 2]
```diff
@@ -1,11 +1,37 @@
...
+            selected_html = mark_safe(' selected="selected"')
+            if not self.allow_multiple_selected:
+                selected_choices.remove(option_value)
+        else:
+            selected_html = ''
+        css_class = ''
+        choice_dict = {str(c.pk): c for c in self.choices.queryset}
+        if option_value in choice_dict.keys() and not choice_dict[option_value].needs_verification:
+            css_class = 'class=no-verification'
+        return format_html('<option {} value="{}"{}>{}</option>',
```
Comment (shaneallgeier) [inline exdb/forms.py]: You could allow for multiple classes by changing this format string to `<option class="{}" value="{}"{}>{}</option>` and then storing space-separated classes in the `css_classes` variable.

File: exdb/forms.py [hunk 3]
```diff
@@ -1,11 +1,37 @@
...
+        option_value = force_text(option_value)
+        if option_value in selected_choices:
+            selected_html = mark_safe(' selected="selected"')
+            if not self.allow_multiple_selected:
+                selected_choices.remove(option_value)
+        else:
+            selected_html = ''
+        css_class = ''
+        choice_dict = {str(c.pk): c for c in self.choices.queryset}
+        if option_value in choice_dict.keys() and not choice_dict[option_value].needs_verification:
```
Comment (shaneallgeier) [inline exdb/forms.py]: There's no need to use the `.keys()` method when checking for `option_value`'s membership in the dict.

File: exdb/forms.py [hunk 4]
```diff
@@ -90,6 +119,11 @@ def clean(self):
         if ex_type.needs_verification and self.cleaned_data.get('attendance'):
             raise ValidationError(_('%(name)s events cannot have an attendance') % {'name': ex_type.name})
 
+        # There are too many branches in this function; this is fixed on
+        # the approval validation branch.
+        if not ex_type.needs_verification and not self.cleaned_data.get('conclusion'):
+            raise ValidationError(_('%(name)s events must have a conclusion') % {'name': ex_type.name})
+
```
Comment (shaneallgeier) [inline exdb/forms.py]: You should check for people who pick a Type that needs verification and accidentally include a value for conclusion. Just set it to the empty string if that happens.

File: exdb/tests/browser_tests.py [hunk 5]
```diff
@@ -324,3 +313,46 @@ class HallStaffDashboardBrowserTest(DefaultLiveServerTestCase):
...
+        attnd_element = self.driver.find_element(By.ID, 'id_attendance')
+        self.assertFalse(attnd_element.find_element(By.XPATH, '..').is_displayed(),
+                         'Attendance field should be hidden when spontaneous is not selected.')
+
+    def test_attendance_conclusion_not_hidden_if_no_verify(self):
+        self.create_spontaneous_type()
+        self.client.get(reverse('create_experience'))
+        type_element = self.driver.find_element(By.ID, 'id_type')
+        type_element.find_element_by_class_name('no-verification').click()
+        type_element.find_elements_by_tag_name('option')[1].click()
```
Comment (shaneallgeier) [inline exdb/tests/browser_tests.py]: This line is unnecessary.

File: exdb/tests/browser_tests.py [hunk 6]
```diff
@@ -324,3 +313,46 @@ class HallStaffDashboardBrowserTest(DefaultLiveServerTestCase):
     def test_load(self):
         self.client.get(reverse('hallstaff_dash'))
         self.assertEqual(self.driver.find_element(By.XPATH, '//h1').text, _('Experiences Pending Approval'))
+
+
+class CreateExperienceBrowserTest(DefaultLiveServerTestCase):
+
+    def create_spontaneous_type(self):
```
Comment (shaneallgeier) [inline exdb/tests/browser_tests.py]: Rename this to `setUp` and then you won't need to call it in all the `test_*` methods.

File: exdb/views.py [hunk 7]
```diff
@@ -193,8 +193,11 @@ def get_success_url(self):
         return reverse('ra_home')
 
     def get_queryset(self):
-        return Experience.objects.filter(Q(author=self.request.user) | (Q(planners=self.request.user) & ~Q(status='dr')),
-                                         start_datetime__gt=timezone.now()).exclude(status__in=('ca', 'co')).prefetch_related('comment_set')
+        user_has_editing_privs = Q(author=self.request.user) | (Q(planners=self.request.user) & ~Q(status='dr'))
+        current_status_allows_edits = ~Q(status__in=('ca', 'co'))
+        event_already_occured = Q(status='ad') & Q(start_datetime__lte=timezone.now())
```
Comment (shaneallgeier) [inline exdb/views.py]: occurred

**Comment by aaaustin10** [comment]
I approve.

File: exdb/tests/integration_tests.py [hunk 8]
```diff
@@ -279,6 +279,15 @@ def test_valid_past_experience_creation_view_submit(self):
...
                          "Experience should have been saved with completed status")
 
+    def test_conclusion_set_to_empty_string_if_needs_verification(self):
+        start = now() + timedelta(days=1)
+        end = now() + timedelta(days=2)
+        data = self.get_post_data(start, end)
+        data['conclusion'] = "Test Conclusion"
+        self.clients['ra'].post(reverse('create_experience'), data)
+        self.assertEqual(Experience.objects.get(name='test').conclusion, "",
+                         "Experience should have been saved with pending status")
```
Comment (shaneallgeier) [inline exdb/tests/integration_tests.py]: This error message should be changed to something related to this test.

**Comment by shaneallgeier** [comment]
I approve.

**Comment by aaaustin10** [comment]
I approve.

**Pull request merged**